### PR TITLE
Add E2E diagnostics support for AMQP and MQTT messaging

### PR DESCRIPTION
**Pull request by erich-wang**
end to end diagnostic design spec for device SDKs is available at  [here](
https://microsoft.sharepoint.com/:w:/r/teams/Azure_IoT/_layouts/15/Doc.aspx?sourcedoc=%7B2B696832-DC1D-4CA1-AF2F-C15647D34146%7D&file=Azure%20IoT%20Hub%20Device%20Client%20SDKs%20-%20E2E%20Diagnostic%20P0%20scenario.docx)

### CODE REVIEW

**Comment by prmathur-microsoft** [comment]
Hi @erich-wang @michaeljqzq  - Can you please send me the spec for these changes ? I need to see the spec to effectively review these changes.

**Comment by SLdragon** [comment]
Hi, @prmathur-microsoft,
The spec document can be found [here](https://microsoft.sharepoint.com/:w:/r/teams/Azure_IoT/_layouts/15/Doc.aspx?sourcedoc=%7B2B696832-DC1D-4CA1-AF2F-C15647D34146%7D&file=Azure%20IoT%20Hub%20Device%20Client%20SDKs%20-%20E2E%20Diagnostic%20P0%20scenario.docx)
If you have any trouble to access this file, please let me know, I will send it to you email.

File: device/iot-device-client/devdoc/requirement_docs/com/microsoft/azure/iothub/diagnosticpropertydata_requirements.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+A message of diagnostic data
+
+## References
+
+## Exposed API
+
+```java
+public class DiagnosticPropertyData
+{
+    public DiagnosticPropertyData(String diagnosticId, String diagnosticCreationTimeUtc);
```
Comment (prmathur-microsoft) [inline device/iot-device-client/devdoc/requirement_docs/com/microsoft/azure/iothub/diagnosticpropertydata_requirements.md]: Why is creation date input ? shouldn't it be the current date/time ?

File: device/iot-device-client/devdoc/requirement_docs/com/microsoft/azure/iothub/diagnosticpropertydata_requirements.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+```java
+public String getDiagnosticId();
+```
+
+**SRS_DIAGNOSTICPROPERTYDATA_01_003: [**The function shall return the value of diagnostic id.**]**
+
+### setDiagnosticId
+
+```java
+public void setDiagnosticId(String diagnosticId);
```
Comment (prmathur-microsoft) [inline device/iot-device-client/devdoc/requirement_docs/com/microsoft/azure/iothub/diagnosticpropertydata_requirements.md]: Why do you need this setter isn't it already provided in constructor ? can diagnostic ID change over time ?

File: device/iot-device-client/devdoc/requirement_docs/com/microsoft/azure/iothub/diagnosticpropertydata_requirements.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+### getDiagnosticCreationTimeUtc
+
+```java
+public String getDiagnosticCreationTimeUtc();
+```
+
+**SRS_DIAGNOSTICPROPERTYDATA_01_005: [**The function shall return the value of diagnostic creation time.**]**
+
+### setDiagnosticCreationTimeUtc
```
Comment (prmathur-microsoft) [inline device/iot-device-client/devdoc/requirement_docs/com/microsoft/azure/iothub/diagnosticpropertydata_requirements.md]: I don't under this setter. Why do you need it ?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java [hunk 4]
```diff
@@ -100,9 +100,11 @@
     private static final String SET_SEND_INTERVAL = "SetSendInterval";
     private static final String SET_CERTIFICATE_PATH = "SetCertificatePath";
     private static final String SET_SAS_TOKEN_EXPIRY_TIME = "SetSASTokenExpiryTime";
+    private static final String SET_DIAGNOSTIC_SAMPLING_PERCENTAGE = "SetDiagnosticSamplingPercentage";
```
Comment (prmathur-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java]: Why can't we add a new API. Why is this going via set option ?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClientDiagnostic.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        this.diagSamplingPercentage = 0;
+        // Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_002: [This constructor shall set message number to 0.]
+        this.currentMessageNumber = 0;
+    }
+
+    public int getDiagSamplingPercentage() {
+        return diagSamplingPercentage;
+    }
+
+    public void setDiagSamplingPercentage(int diagSamplingPercentage) {
```
Comment (prmathur-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClientDiagnostic.java]: Why do you need a setter ? why can't constructor take it ? isn't it mandatory ?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClientDiagnostic.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+            return (char) ('a' + value - DIAGNOSTIC_ID_CHARACTER_CAPITALIZED - DIAGNOSTIC_ID_CHARACTER_NUMBER);
+        }
+    }
+
+    private String generateEightRandomCharacters()
+    {
+        // Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_005: [This function shall generate 8 random chars, each is from 0-9a-z.]
+        StringBuilder result = new StringBuilder();
+        for (int i = 0; i < 8; i++) {
+            int randomNum = ThreadLocalRandom.current().nextInt(0, DIAGNOSTIC_ID_CHARACTER_BASE);
```
Comment (prmathur-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClientDiagnostic.java]: use Random class to generate this

Comment (prmathur-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClientDiagnostic.java]: or UUID if its within your restrictions.

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClientDiagnostic.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    {
+        boolean result = false;
+        // Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_007: [This function shall return false if sampling percentage is set to 0.]
+        if (diagSamplingPercentage > 0 && diagSamplingPercentage <= 100) {
+            if (currentMessageNumber == Integer.MAX_VALUE) {
+                currentMessageNumber = 0;
+            }
+            currentMessageNumber++;
+            // Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_008: [This function shall return value due to the sampling percentage setting.]
+            result = (Math.floor((currentMessageNumber - 2) * diagSamplingPercentage / 100.0) < Math.floor((currentMessageNumber - 1) * diagSamplingPercentage / 100.0));
```
Comment (prmathur-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClientDiagnostic.java]: I can't seem to follow this math your are doing to determine if you should add diagnostic info. Why isn't it as simple as if sampling rate is set then you add diagnostic info and no if not ?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/Message.java [hunk 8]
```diff
@@ -133,6 +133,9 @@
     private ByteArrayInputStream bodyStream;
     private CustomLogger logger;
 
+    // Diagnostic data attached to the message
+    private DiagnosticPropertyData diagnosticPropertyData;
```
Comment (prmathur-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/Message.java]: Why do you need a separate class for adding two properties ?

**Comment by tameraw** [comment]
Closing.

**Pull request closed without merging**